contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,204 @@
+
+It uses the syntax `module.export` and `require`.
+
+`anExampleModule.js`
+```js
+var variableOne = 1;
+var variableTwo = 2;
+
+getVariableOne = function() {
+  return variableOne;
+}
+
+sumOfVariable = function() {
+    return variableOne + variableTwo;
+}
+
+module.exports = {
+  getVariableOne: getVariableOne,
+  sumOfVariable: sumOfVariable,
+};
+```
+
+`index.js`
+```js
+var anExampleModule = require('./anExampleModule.js');
+
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### EcmaScript2015 / ES6 modules
+
+In 2015, JavaScript introduced built-in modules in ES6. It uses the syntax `import` and `export`.
+
+`anExampleModule.js`
+```js
+var variableOne = 1;
+var variableTwo = 2;
+
+export getVariableOne() {
+    return variableOne;
+}
+
+export sumOfVariable() {
+    return variableOne + variableTwo;
+}
+```
+
+`index.js`
+```js
+import * as anExampleModule from './anExampleModule.js';
+
+anExampleModule.getVariableOne(); // 1
+anExampleModule.sumOfVariable(); // 3
+```
+
+### Which to use?
+
+While ES6 modules is officially the most recent JavaScript standard rolled out, it is still not widely supported by all browsers. It needs transpiler <sup>[link](https://scotch.io/tutorials/javascript-transpilers-what-they-are-why-we-need-them)</sup> such as Babel <sup>[link](https://babeljs.io/)</sup> for unsupported browser. Similar to CommonJS, ES6 modules also requires bundler such as Webpack.
+
+Despite ES6 modules not widely supported yet, it supports various advanced features not available in CommonJS such as asynchronous loading, support for tree shaking, static code analysis, etc. These features will not be covered for now.
+
+Nevertheless, both ES6 and CommonJS are a very valid approaches in modularizing your JavaScript project. You can take this opportunity to learn transpiling and bundling of JavaScript.
+
+The invention of ES6 and CommonJS has not made the old module pattern obsolete. It is still a valid approach in modularizing your code. It has the advantage of not requiring transpiling and bundling of your JavaScript code.
+
+All in all, the best approach depends on your project requirements.
+
+## How to start?
+
+You can start with module pattern right away by refactoring segments of your code into different files and wrapping the code in IIFE.
+
+If you are planning to use CommonJS or ES6, you can start by refactoring segments of your code, but you will also need to be familiar with transpiler and bundler such as Babel and Webpack.
+
+## Further Reading
+
+You can read more on JavaScript modules at following websites:
+
+[JavaScript Modules: A Beginner’s Guide](https://medium.freecodecamp.org/javascript-modules-a-beginner-s-guide-783f7d7a5fcc) <br>
```

Comment:
It's good to give your own take on these resources. Or describe what value they can provide